### PR TITLE
metrics: implement HTTP rate limiting - try#2

### LINKED ISSUE #2012: VM Metric aggregator http rate-limiting
/kind enhancement

#1840 introduces an aggregated VMI metrics collector endpoint into the virt-handler component. This endpoint involves reaching into each of the guest's pods to retrieve information from the guest's corresponding libvirt processes. Depending on how many VMIs are running on a node, this isn't an action we'd want to be performing often. 

We need some basic form of project that ensures an external collector (or collectors) can't inadvertently abuse this collection endpoint by making too frequent of calls. This also protects us from anyone who may attempt to profile the VMS using this collection endpoint at too frequent of an interval.

### PR DESCRIPTION
**Pull request by fromanirh**
**What this PR does / why we need it:**
This is another attempt to fix #2012 . Content is the same of PR https://github.com/kubevirt/kubevirt/pull/2082 but fixes are squashed together, and here I avoid unneeded changes.
The main driver behind this concurrent PR is try to reduce the mess I made in 2082, where I feel I did something wrong to bazel, there are too many (conflicting) changes to bazel.build files.
Here the changes seems simpler and more logical (from my limited unterstanding).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubevirt/kubevirt/issues/2012

**Special notes for your reviewer**:
See discussion in the issue for the rationale

**Release note**:

```release-note
virt-handler now implement a tunable to limit the maximum concurrent requests to the metrics endpoint
```

### CODE REVIEW

**Comment by rmohr** [comment]
> The main driver behind this concurrent PR is try to reduce the mess I made in 2082, where I feel I did something wrong to bazel, there are too many (conflicting) changes to bazel.build files.

On a rebase, if you didn't explicitly change anything manually in the bazel build files (which you would normally only do when adding new binaries), you can always accept the buildfile changes from master and simply run `make generate` afterwards again.

**Comment by fromanirh** [comment]
> > The main driver behind this concurrent PR is try to reduce the mess I made in 2082, where I feel I did something wrong to bazel, there are too many (conflicting) changes to bazel.build files.
> 
> On a rebase, if you didn't explicitly change anything manually in the bazel build files (which you would normally only do when adding new binaries), you can always accept the buildfile changes from master and simply run `make generate` afterwards again.

Understood. I actually changed pkg/monitoring/vms/prometheus/BUILD.bazel manually, because without this line
```
"//vendor/github.com/prometheus/client_golang/prometheus/promhttp:go_default_library",
```

`make functest` failed locally.

@rmohr do you want to keep this PR or https://github.com/kubevirt/kubevirt/pull/2082 ? I'm fine with whatever you maintainers prefer.

**Comment by rmohr** [comment]
> `make functest` failed locally.

`make generate` adds that line for you, you don't have to do this. Only explicit rules need to be added manually.

**Comment by fromanirh** [comment]
> > `make functest` failed locally.
> 
> `make generate` adds that line for you, you don't have to do this. Only explicit rules need to be added manually.

I see. Now everything makes sense: I did `make deps-update`, but I did not `make generate`. Lesson learned :)

**Comment by fromanirh** [comment]
@rmohr sure. Been busy with other PRs. Will update ASAP.

**Comment by fromanirh** [comment]
@rmohr test added, but I'm not quite happy about it, it looks quite fragile. I'd like to avoid flaky tests. The issue here is: it is hard to trigger _reliably_ a stressfull enough load to let the throttling kick in. Injecting delays in the monitoring code to simulate high load would make it much better, but is this an option?

File: tests/infra_test.go [hunk 1]
```diff
@@ -104,6 +104,57 @@ var _ = Describe("Infrastructure", func() {
...
+							"-k",
+							fmt.Sprintf("https://%s:%s/metrics", pod.Status.PodIP, "8443"),
+						})
+					errors <- err
+				}()
+			}
+
+			errorCount := 0
+			for ix := 0; ix < concurrency; ix++ {
+				err := <-errors
```
Comment (fromanirh) [inline tests/infra_test.go]: TODO: add a comment to explain *which* error are we expecting. E.g. 50x HTTP code? timeout?

**Comment by rmohr** [comment]
> @rmohr test added, but I'm not quite happy about it, it looks quite fragile. I'd like to avoid flaky tests. The issue here is: it is hard to trigger _reliably_ a stressfull enough load to let the throttling kick in. Injecting delays in the monitoring code to simulate high load would make it much better, but is this an option?

netcat could do the job: https://stackoverflow.com/questions/14339111/can-curl-be-used-to-mimic-a-persistent-but-idle-open-connection

**Comment by fromanirh** [comment]
@rmohr thanks for the hints. I'll update the PR accordingly ASAP.

File: tests/infra_test.go (hunk 1 above)
Comment (rmohr) [reply tests/infra_test.go]: What do we expect?

Comment (fromanirh) [reply tests/infra_test.go]: We expect _any_ error to be occurred. Here I'm collecting the statuses from the concurrent clients - this loop is also meant to act as sync point (will document with a comment)

**Comment by fromanirh** [comment]
Work on this PR is been delayed by other task, but I'm totally not abandoning it.

**Comment by fromanirh** [comment]
> Work on this PR is been delayed by other tasks, but I'm totally not abandoning it.

Updated, at last! Had to force-push changes because seriously out of date (e.g. transition to go.mod), let's see how CI goes.

**Pull request merged**